uber/site_sections/budget.py

```diff
@@ -139,44 +142,89 @@ def generate_promo_codes(
+            'uses_allowed': uses_allowed
+        }
+
         if cherrypy.request.method == 'POST':
             codes = None
             if is_single_promo_code:
-                if params.get('code', '').strip():
-                    codes = [params['code']]
-                else:
-                    count = 1
+                count = 1
+                if code:
+                    codes = [code]
 
             if use_words and not codes and \
                     not any(s for (_, s) in words.items()):
-                raise HTTPRedirect('generate_promo_codes?message='
-                    'Please add some promo code words!')
+                result['message'] = 'Please add some promo code words!'
+                return result
 
             if not codes:
                 if use_words:
                     codes = PromoCode.generate_word_code(count)
                 else:
-                    length = int(params.get('length', 12))
-                    segment_length = int(params.get('segment_length', 3))
+                    try:
+                        length = int(length)
+                    except:
+                        length = 9
+                    try:
+                        segment_length = int(segment_length)
+                    except:
+                        segment_length = 3
                     codes = PromoCode.generate_random_code(
                         count, length, segment_length)
 
             promo_codes = []
             for code in codes:
-                params['code'] = code
-                promo_codes.append(PromoCode().apply(params, restricted=False))
+                promo_codes.append(PromoCode().apply({
+                        'code': code,
+                        'expiration_date': expiration_date,
+                        'discount_type': discount_type,
+                        'discount': discount,
+                        'uses_allowed': uses_allowed,
+                    }, restricted=False))
 
             message = check_all(promo_codes)
             if message:
-                raise HTTPRedirect('generate_promo_codes?message={}', message)
+                result['message'] = message
+                return result
 
             result['promo_codes'] = session.bulk_insert(promo_codes)
-            if len(result['promo_codes']) != count:
+            generated_count = len(result['promo_codes'])
+            if generated_count <= 0:
+                result['message'] = "Could not generate any of the " \
+                    "requested promo codes. Perhaps they've all been taken " \
+                    "already?"
+                return result
+
+            if generated_count != count:
                 result['message'] = 'Some of the requested promo codes ' \
                     'could not be generated'
 
         if 'export' in params:
             return self.export_promo_codes(codes=result['promo_codes'])
+
+        result['is_single_promo_code'] = 1
```

Comment:
If there's ever any validation failures we return early. If it makes it all the way to the end of the function, then it was successful, and at that point we DO reset to the default values.

I was half distracted while writing this, so there's probably a better way to do it.